rfd/xxxx-bot-for-cert-renewals.md

```diff
@@ -0,0 +1,627 @@
+certificate would be otherwise useless for end users as it would not provide 
+access to any resources.
+
+Next, users separately configure one or more certificate outputs. As part of
+each renewal event, the bot generates an additional set of *non-*-renewable
+certificates and stores them in the designated output. Each of these outputs
+may optionally be scoped to a subset of all roles the bot is capable of
+assuming.
+
+This design is meant to encourage deployments where the bot's own data
+directory, containing the renewable certificate, is separate and inaccessible
+to other users on the system, via the native security methods of the output
+sink (e.g. Unix permissions, Kubernetes RBAC, etc). The secondary certificates
+may be written with relaxed permissions to enable use by end users. If these
+secondary certificates are stolen, not only will they not be renewable, but may
+also have only a subset of the permissions granted to the bot.
+
+Unfortunately, Teleport today does not fully support the impersonation UX we
+would like to provide with the bot. As is, impersonation requires two `User`
+resources: one for the impersonator, and one for the impersonatee. Today, we
+could require that users manually create a Teleport user for each set of roles
+the bot may assume, however this UX is undesirable. Therefore, we propose
+adding the ability to, when requesting user certificates, allow the bot to
+"impersonate" its own `User`, but with a set of additional roles as specified
+in its certificate. This would enable use of the previously-described
+`tctl bots add --roles=...` UX. If desired, this UX remains compatible with
+other-user impersonation should we wish to add an additional `--users=...`
+flag.
+
+#### Renewals
+
+There are several scenarios under which the bot will initiate a renewal:
+
+1. When a certificate is nearing expiry. A certificate is considered near
+   expiration if 75% of its TTL has elapsed, or when there are 4 hours or less
+   until expiration (whichever is sooner).
+2. When a user and/or host CA rotation is taking place.
+3. When a renewal is requested manually. For testing and debugging purposes, the
+   bot will expose an API endpoint that can be used to trigger a renewal. This
+   API will be accessible on the loopback interface only.
+
+While the bot generates multiple sets of certificates - that is, one primary
+renewable certificate and many secondary end-user certificates - they are all
+renewed at roughly the same time, and secondary certificates have an expiration
+matching the primary certificate. It may be possible in the future to have
+secondary certificates with a shorter TTL than the primary (probably 1/n
+durations) but this is out of scope for the initial implementation.
+
+*TODO (Tim):* I think the renewal interval should be some fraction of the total
+TTL (at most 1/2) rather than 75%. For example, given a 4 hour TTL, we should
```

Comment:
I don't disagree.  The 75% was a totally arbitrary number I came up with, so I would just propose whatever you want for the group to discuss.

Reply:
Fair enough, I'll make the change.